modules/cnn_3dobj/include/opencv2/cnn_3dobj.hpp

```diff
@@ -0,0 +1,252 @@
+        @param cols Columns of a single sample image.
+        @param headerPath Path where the header will be stored.
+         */
+        CV_WRAP static void createHeader(int num_item, int rows, int cols, const char* headerPath);
+        /** @brief Write binary files used for training in other open source project including Caffe.
+        @param filenameImg Path which including a set of images.
+        @param binaryPath Path which will output a binary file.
+        @param headerPath Path which header belongs to.
+        @param num_item Number of samples.
+        @param label_class Class label of the sample.
+        @param x Pose label of X.
+        @param y Pose label of Y.
+        @param z Pose label of Z.
+        @param isrgb Option for choice of using RGB images or not.
+         */
+        CV_WRAP static void writeBinaryfile(string filenameImg, const char* binaryPath, const char* headerPath, int num_item, int label_class, int x, int y, int z, int isrgb);
+    };
+
+/** @brief Caffe based 3D images descriptor.
+ The class extract feature which is discriminant for class and pose from given samples @cite wohlhart15.
+ */
+
+/************************************ Feature Extraction Class ************************************/
+    class CV_EXPORTS_W descriptorExtractor
+    {
+        private:
+        caffe::Net<float>* convnet;
+        cv::Size input_geometry;
+        int num_channels;
+        bool net_set;
+        int net_ready;
+        cv::Mat mean_;
+        std::vector<string> device_info;
+        /** @brief Load the mean file in binaryproto format if it is needed.
+        @param mean_file Path of mean file which stores the mean of training images, it is usually generated by Caffe tool.
+         */
+        void setMean(const string& mean_file);
+        /** @brief Wrap the input layer of the network in separate cv::Mat objects(one per channel).
+         This way we save one memcpy operation and we don't need to rely on cudaMemcpy2D.
+         The last preprocessing operation will write the separate channels directly to the input layer.
+         */
+        void wrapInput(std::vector<cv::Mat>* input_channels);
+        /** @brief Convert the input image to the input image format of the network.
+         */
+        void preprocess(const cv::Mat& img, std::vector<cv::Mat>* input_channels);
+        public:
+        /** @brief Set the device for feature extraction, if the GPU is used, there should be a device_id.
+        @param device_type CPU or GPU.
+        @param device_id ID of GPU.
+         */
+        descriptorExtractor(const string& device_type, int device_id);
+        /** @brief Get device information for feature extraction.
+         */
+        std::vector<string> getDevice();
+        /** @brief Set device information for feature extraction, if the descriptor is already loaded and
+         you want to change the device for featrue extraction, this function is convenient.
```

Comment:
replace:
Set device information for feature extraction. Useful to change device without the need to reload the net